igor-web/src/main/java/com/netflix/spinnaker/igor/helm/HelmMonitor.java

```diff
@@ -0,0 +1,207 @@
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.igor.helm;
+
+import static net.logstash.logback.argument.StructuredArguments.kv;
+
+import com.netflix.spectator.api.Registry;
+import com.netflix.spinnaker.igor.IgorConfigurationProperties;
+import com.netflix.spinnaker.igor.build.model.GenericArtifact;
+import com.netflix.spinnaker.igor.config.HelmProperties;
+import com.netflix.spinnaker.igor.helm.accounts.HelmAccount;
+import com.netflix.spinnaker.igor.helm.accounts.HelmAccounts;
+import com.netflix.spinnaker.igor.helm.cache.HelmCache;
+import com.netflix.spinnaker.igor.helm.model.HelmIndex;
+import com.netflix.spinnaker.igor.history.EchoService;
+import com.netflix.spinnaker.igor.history.model.HelmEvent;
+import com.netflix.spinnaker.igor.polling.*;
+import com.netflix.spinnaker.kork.discovery.DiscoveryStatusListener;
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService;
+import java.util.*;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+
+@Service
+@ConditionalOnProperty("helm.enabled")
+public class HelmMonitor
+    extends CommonPollingMonitor<HelmMonitor.HelmDelta, HelmMonitor.HelmPollingDelta> {
+  private final HelmCache cache;
+  private final HelmAccounts helmAccounts;
+  private final HelmProperties helmProperties;
+  private final Optional<EchoService> echoService;
+
+  @Autowired
+  public HelmMonitor(
+      IgorConfigurationProperties properties,
+      Registry registry,
+      DynamicConfigService dynamicConfigService,
+      DiscoveryStatusListener discoveryStatusListener,
+      Optional<LockService> lockService,
+      HelmCache cache,
+      HelmAccounts helmAccounts,
+      HelmProperties helmProperties,
+      Optional<EchoService> echoService,
+      TaskScheduler scheduler) {
+    super(
+        properties,
+        registry,
+        dynamicConfigService,
+        discoveryStatusListener,
+        lockService,
+        scheduler);
+    this.cache = cache;
+    this.helmAccounts = helmAccounts;
+    this.helmProperties = helmProperties;
+    this.echoService = echoService;
+  }
+
+  @Override
+  protected void initialize() {}
+
+  @Override
+  public void poll(boolean sendEvents) {
+    helmAccounts.updateAccounts();
+    helmAccounts.accounts.forEach(
+        account -> pollSingle(new PollContext(account.name, account.toMap(), !sendEvents)));
+  }
+
+  @Override
+  public PollContext getPollContext(String partition) {
+    Optional<HelmAccount> account =
+        helmAccounts.accounts.stream().filter(it -> it.name.equals(partition)).findFirst();
+    if (!account.isPresent()) {
+      throw new IllegalStateException(
```

Comment:
Please use the `SpinnakerException` family.

Reply:
Will change to `ConstraintViolationException`